Allows generation of just a subset of the diagrams by passing command line options.

gspencergoog opened the pull request:
Passing either --category or --name to generate.dart will restrict the generated diagrams to just those diagrams that match.  The flags can be specified multiple times to specify more than one.  The category is the directory name that the assets end up in, and the name can be the base name of the diagram, with or without the suffix.

Hixie commented:
Did you intend the formatting changes to be included? (It's fine if you did, happy to review them, but there have been past PRs where you indicate that was accidental so I thought I'd check before reviewing.)

gspencergoog commented:
Thanks, no, I didn't intend all of the formatting changes: I had changed my line length for dartfmt to 80 chars while I was working on the dartdoc change, and apparently that isn't a project-specific setting.  I've reverted those.

Hixie commented on diagrams/lib/main.dart:
```diff
@@ -26,7 +27,8 @@ import 'tile_mode.dart';
   final Directory directory = new Directory(
-    path.join((await getApplicationDocumentsDirectory()).absolute.path, 'diagrams'),
+    path.join(
+        (await getApplicationDocumentsDirectory()).absolute.path, 'diagrams'),
```
missing trailing comma for better styling

gspencergoog replied:
Fixed.

Hixie commented on diagrams/lib/main.dart:
```diff
@@ -35,7 +37,20 @@ Future<Directory> prepareOutputDirectory() async {
+      : <String>[];
+  final ArgParser parser = new ArgParser();
+  parser.addMultiOption('category',
+      help: 'Only create the diagrams for the given category.');
```
not sure we really need the help lines here, they're redundant with the "real" ones in the main file

gspencergoog replied:
Yeah, I guess so: you can't really see them anyhow.

Hixie commented on diagrams/lib/tile_mode.dart:
```diff
@@ -129,8 +129,11 @@ class TileModeDiagramStep extends DiagramStep {
     for (TileModeDiagram tileModeDiagram in tileModeDiagrams) {
+      if (onlyGenerate != null && !onlyGenerate.contains(tileModeDiagram.name)) {
+        continue;
+      }
```
it's a pity we have to duplicate this pattern everywhere. not sure how we can avoid it... maybe just filter by `category` instead, rather than by any specific filename? That way you could factor the check out into the superclass...

gspencergoog replied:
There already is the ability to filter by category (see `--category` on generate.dart). This just refines it (or provides a separate filter) for filenames.  I thought about making a separate base class that represents a single diagram and has a name (maybe Diagram?).  Each DiagramStep can then provide a list of Diagrams from an accessor, and the base generateDiagrams would call generateDiagram for each one where the filename matches, instead of subclasses overriding generateDiagrams.  I'll see if that works well.

gspencergoog replied:
OK, I refactored it so that the filtering happens in the step.  It does seem to make things a bit cleaner for the implementations.

Hixie commented:
![LGTM](http://junkyard.damowmow.com/579)

gspencergoog commented:
Sorry, I upended things quite a bit when I refactored to move the filtering into the base DiagramStep.  Could you take another look?

I also got rid of the filename generator callback, since it was really not being used, and was redundant with the name parameter to the animation creation routines.

Hixie commented on diagrams/generate.dart:
```diff
@@ -277,10 +277,18 @@ class DiagramGenerator {
+  static String get assetDir {
+    return path.joinAll(path.split(projectDir)
+      ..removeLast()
+      ..add('assets'));
```
there's some misleading indentation here. Looks like the cascade applies to the return value of `path.joinAll`, but it's actually applying to the argument.

gspencergoog replied:
Fixed.

Hixie commented on diagrams/generate.dart:
```diff
@@ -469,5 +493,6 @@ Future<Null> main(List<String> arguments) async {
   }
 
-  new DiagramGenerator(temporaryDirectory: temporaryDirectory, cleanup: !keepTmp)..generateDiagrams();
+  new DiagramGenerator(temporaryDirectory: temporaryDirectory, cleanup: !keepTmp)
```
we should rename "keepTmp" to "keepTemporaryDirectory" while we're at it

gspencergoog replied:
Done

gspencergoog commented on diagrams/lib/app_bar.dart:
```diff
@@ -114,11 +117,12 @@ class AppBarDiagramStep extends DiagramStep {
+  Future<File> generateDiagram(DiagramMetadata diagram) async {
+    final AppBarDiagram typedDiagram = diagram;
+    controller.builder = (BuildContext context) => typedDiagram;
+    return await controller.drawDiagramToFile(new File('${diagram.name}.png'));
```
Yeah, I noticed that too.  I thought about maybe just passing the widget to `drawDiagramToFile` and letting the controller make its own builder, but for animated diagrams it seems to make sense to keep the builder pattern. Ultimately, I decided I had done enough refactoring in this PR already. :-)
